_posts/2018-11-10-storage-in-ethereum.md

```diff
@@ -8,52 +8,56 @@ categories: ethereum
+
+If we use an in-memory EVM cache, when should we commit these modifications from the cache to storage? To answer this question, we need to look at the transaction receipt. The transaction receipt is a data structure that contains information about the executed transaction (used gas etc). It also contains the state root hash that exists immediately after transaction execution (for pre-Byzantium blocks). So we have to commit changes after we receive the transaction receipt (which contains the state root hash from the Merkle Patricia trie storage). 
 
-When should we commit modifications stored in the EVM cache? To answer this question we have to understand what transaction receipt is. A transaction receipt is a data structure that contains some information about the executed transaction (used gas etc). It also contains state root hash right after transaction execution (for pre-Byzantium blocks). So as you can see we have to commit changes after transaction execution to get state root hash from the Merkle Patricia trie storage. But this approach has one problem.
+Unfortunately, there is a problem with this approach. 
 
 ![evm-cache](https://i.imgur.com/ZIFq2NB.png)
 
-### Cache storage for block
+### Cache storage for blocks
 
-The problem with the previous approach is that we should commit cache after every transaction. It's ok if we'll be importing only valid blocks. But in the real network, some peers will share invalid blocks and we shouldn't commit changes made by their transactions.
+The problem with the previous approach is that we should commit the cache after every transaction. That works if we're only importing valid blocks. But in the real network, some peers will share invalid blocks and we shouldn't commit changes made in those transactions.
 
-We can introduce a second level of cache. It will be in-memory trie where transaction changes will be committed from the EVM cache. It will solve the problem with invalid block data because we will be able to simply discard in-memory tries from these blocks.
+To fix this, we introduce a second cacheing level. We can use an in-memory trie where transaction changes are committed from the EVM cache. It solves the invalid data problem because we simply discard in-memory tries from invalid blocks.
 
-In-memory trie will read not existing in-memory nodes from permanent disk storage but it will write all changes to memory. So another optimization we will get from in-memory trie is that update/read speed is much faster from memory than disk.
+An in-memory trie will read non-existing in-memory nodes from permanent disk storage and it will write all changes to memory. Another optimization we get from the in-memory trie is that update/read speed is much faster from memory than from disk.
```

Comment:
I don't totally understand this sentence. What do you mean by 'it will read not existing in-memory nodes from permanent disk storage? Guess I'm confused by the non-existing or not existing wording.